In [1]:
%load_ext rpy2.ipython

/usr/local/lib/python3.6/dist-packages/rpy2/robjects/pandas2ri.py:14: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.
  from pandas.core.index import Index as PandasIndex
/usr/local/lib/python3.6/dist-packages/rpy2/robjects/pandas2ri.py:34: UserWarning: pandas >= 1.0 is not supported.
  warnings.warn('pandas >= 1.0 is not supported.')


### В этом уроке познакомимся с синтаксисом циклов в R.

In [2]:
!wget  https://stepic.org/media/attachments/lesson/11481/evals.csv.

--2020-04-06 18:20:59--  https://stepic.org/media/attachments/lesson/11481/evals.csv.
Resolving stepic.org (stepic.org)... 13.69.75.239
Connecting to stepic.org (stepic.org)|13.69.75.239|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2020-04-06 18:20:59 ERROR 404: Not Found.



Создайте новую числовую переменную  new_var в данных mtcars, которая содержит единицы в строчках, если в машине не меньше четырёх карбюраторов (переменная "carb") или больше шести цилиндров (переменная "cyl"). В строчках, в которых условие не выполняется, должны стоять нули.

In [0]:
%%R 
mtcars$new_var  <- ifelse((mtcars$carb == 4) | (mtcars$cyl <= 4), 1, 0)

В уже существующей переменной my_vector сохранен вектор из 50 чисел.

Решите задачу используя конструкцию:
```
if () {

} else {

} 
```
Если среднее значение вектора my_vector больше 20, в переменную result сохраните "My mean is great",  если среднее значение my_vector меньше или равно 20 то в переменную result сохраните  строку "My mean is not so great".

In [0]:
%%R 
if (mean(my_vector) > 20) { result <- "My mean is great"

} else { result <- "My mean is not so great"

} 

В этой задаче от вас потребуется узнать некоторую информацию о типах данных в R самостоятельно! Встроенные в R данные AirPassengers - это новый для нас формат данных типа Time-Series. Изучите структуру этих данных, прежде чем начать решение задачи! Например напишите команды:

```
> ?AirPassengers # справка о данных
> str(AirPassengers) # структура данных
```
В встроенных в R данных AirPassengers хранится 144 значения (количество пассажиров в месяц) с 1949 по 1960 год. Данные Time-Series очень похожи на вектор по своей структуре, например мы можем обратиться к любому из 144 элементов используя уже знакомую нам индексацию AirPassengers[1] или AirPassengers[56].

Можно вообще перевести исходные данные в вектор при помощи команды as.vector(AirPassengers) и продолжить с ними работу как с вектором.

И так ваша задача создать переменную good_months и сохранить в нее число пассажиров только в тех месяцах, в которых это число больше, чем показатель в предыдущем месяце.  

Важный момент! В R оператор : для создания последовательности имеет приоритет над арифметическими действиями. Таким образом, если у вас есть переменная i, равная 10, и вы хотите создать вектор от 1 до i - 1, воспользуйтесь скобками, чтобы указать последовательность действий.
```
> i <- 10
> 1 : i - 1 # так мы создадим последовательность от 1 до 10, а потом вычтем единицу из каждого элемента
[1] 0 1 2 3 4 5 6 7 8 9

> 1 : (i - 1) # а вот так мы создадим последовательность от 1 до i - 1, то есть от 1 до 9. 
[1] 1 2 3 4 5 6 7 8 9
```

In [22]:
%%R 
good_months <- c()
for(i in 1:(length(AirPassengers)-1)) {
     if (AirPassengers[i] < AirPassengers[i+1]){ 
     good_months <- c(good_months,AirPassengers[i+1])
   }
   }
good_months

 [1] 118 132 135 148 118 126 141 149 170 140 145 150 178 172 178 199 166 171 180
[20] 193 183 218 230 242 194 196 236 243 264 272 201 204 235 234 264 302 229 242
[39] 267 269 270 315 364 278 284 317 318 374 413 306 315 356 355 422 465 467 336
[58] 340 362 363 435 491 505 337 360 406 420 472 548 559 405 417 419 461 472 535
[77] 622 432


Задачка для супер героев, повышенной сложности!

Для встроенных в R данных AirPassengers рассчитайте скользящее среднее с интервалом сглаживания равным 10. Напечатайте получившийся результат (первым значением в выводе должно быть среднее для элементов 1:10, во втором значении - среднее для элементов 2:11 и т.д., в последнем  - среднее для элементов 135 :144)

Все полученные значения средних сохраните в переменную moving_average.

Пример расчета для вектора из 11 элементов:

![](https://ucarecdn.com/7d59b25c-8c21-44d0-a858-b8d53d37bd1f/)







Соответственно, для наших данных из 144 наблюдений должно получиться 135 средних (первые и последние 4 средних):
```
129.8 129.0 129.0 127.3 ... 483.6 489.2 486.5 490.6
```
Важный момент! В R оператор : для создания последовательности имеет приоритет над арифметическими действиями. Таким образом, если у вас есть переменная i, равная 10, и вы хотите создать вектор от 1 до i - 1, воспользуйтесь скобками, чтобы указать последовательность действий.
```
> i <- 10
> 1 : i - 1 # так мы создадим последовательность от 1 до 10, а потом вычтем единицу из каждого элемента
[1] 0 1 2 3 4 5 6 7 8 9

> 1 : (i - 1) # а вот так мы создадим последовательность от 1 до i - 1, то есть от 1 до 9. 
[1] 1 2 3 4 5 6 7 8 9
```
Если вам потребуется создать вектор moving_average заранее, то есть несколько способов сделать это:
1. самый простой, но не очень правильный вариант - создать пустой вектор
```
moving_average <- c()
```
2. можно сразу создать вектор определенной длины и определенного типа:
```
moving_average <- numeric(135)
```
Такой вариант является более предпочтительным. Почему? Узнаем во второй части курса!)

А для тем, кто уже выбрал путь воина и не хочет использовать цикл - советую познакомиться с функцией cumsum. Подсказка: если у нас есть два вектора одинаковой длинны, то если из одного вектора вычесть второй вектор, мы найдем разность для первых элементов векторов, затем для вторых и т.д.

```
> x <- c(2, 4, 7)

> y <- c(2, 3, 5)

> x - y

[1] 0 1 2
```

In [39]:
%%R 
moving_average <- c()
for(i in 1:(length(AirPassengers)-9)) {
       moving_average <- c(moving_average, mean(AirPassengers[c(i:(i+9))]))
   }
   
moving_average

  [1] 129.8 129.0 129.0 127.3 127.0 129.0 129.0 126.7 126.8 130.2 135.3 140.7
 [13] 142.2 142.1 143.5 143.9 145.4 150.7 152.1 152.3 153.1 157.2 163.8 170.8
 [25] 173.0 173.1 174.7 174.0 175.7 177.8 178.1 176.5 178.4 183.0 191.0 197.3
 [37] 199.8 199.9 201.3 201.6 203.1 208.4 210.1 210.0 210.1 215.6 223.7 230.2
 [49] 231.9 230.3 230.8 227.6 222.9 223.5 221.9 218.9 218.1 224.6 232.8 240.7
 [61] 243.5 243.4 247.5 248.2 248.8 252.1 252.6 249.4 251.6 262.1 273.9 284.8
 [73] 289.3 288.8 293.3 295.0 295.8 300.5 300.3 295.7 298.4 308.5 321.6 333.4
 [85] 336.2 334.9 337.8 337.6 336.4 340.2 337.6 331.8 333.5 344.5 360.6 373.9
 [97] 378.0 377.0 380.5 378.9 375.9 376.6 369.2 359.0 355.8 364.5 380.3 390.2
[109] 392.5 389.5 391.4 391.2 390.6 394.9 391.0 383.9 380.6 395.0 415.0 430.3
[121] 437.3 437.5 443.8 444.9 444.4 444.3 443.2 435.6 433.2 449.1 469.0 483.6
[133] 489.2 486.5 490.6
